sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -215,18 +215,61 @@ class JoinSuite extends QueryTest with SharedSQLContext {
           Row(1, null, 2, 2) ::
           Row(2, 2, 1, null) ::
           Row(2, 2, 2, 2) :: Nil)
+
+      checkAnswer(
+        testData3.as("x").join(testData3.as("y"), $"x.a" > $"y.a"),
+        Row(2, 2, 1, null) :: Nil)
+    }
+    withSQLConf(SQLConf.CROSS_JOINS_ENABLED.key -> "false") {
+      val msg = "Detected cartesian product for INNER join between logical plans"
+      var e = intercept[AnalysisException](testData3.join(testData3).collect())
+      assert(e.getMessage.contains(msg))
+      e = intercept[AnalysisException] {
+        testData3.as("x").join(testData3.as("y"), $"x.a" > $"y.a").collect()
+      }
+      assert(e.getMessage.contains(msg))
+    }
+  }
+
+  test("subquery converted to cartesian product join") {
```

Comment:
This is another typical example for cartesian product, which is converted from subquery.